### PR TITLE
[Op] Fuse bias+dropout in FusedMLP

### PR DESCRIPTION
**Pull request by comaniac**
## Description ##
1. Fuse bias and dropout in op.FusedMLP.
2. Add unit tests for op.FusedMLP.
3. Correct random states in GPT schedules when sequence parallelism is enabled.

## Checklist ##

- [x] PR's title starts with a category (e.g. [Bugfix], [Model], [Tutorial], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

cc @chhzh123 @szhengac

### CODE REVIEW

File: slapo/model_schedule/gpt2.py [hunk 1]
```diff
@@ -478,6 +478,10 @@ def shard_parameters(
                 sub_sch["module.out_proj"].sync(
                     mode="fwd_post", sync_op_or_fn="reduce_scatter", axis=1
                 )
+
+                # In this case, both attention droput and residual dropout
+                # have to use different random seeds.
+                sub_sch["module"].fork_rng()
```
Comment (szhengac) [inline slapo/model_schedule/gpt2.py:484]: with sequence parallel, we do not need to specify fork rng since different TP ranks already use different seeds

File: slapo/op/fused_bias.py [hunk 2]
```diff
@@ -118,3 +109,52 @@ def reset_parameters(self, prev_weight=None):
...
+        self.training = training
+        self.inplace = inplace
+        self.func = torch.jit.script(bias_dropout)
+        # Somehow memory_efficient_fusion generates a different dropout mask
+        # against the original dropout function even the random seed is the same.
+        # if memory_efficient_fusion is not None:
+        #     bias_dropout_func = partial(
+        #         bias_dropout, p=p, training=training, inplace=inplace
+        #     )
+        #     self.func = memory_efficient_fusion(bias_dropout_func)
```
Comment (szhengac) [inline slapo/op/fused_bias.py:149]: what is this memory_efficient_fusion?

File: slapo/op/fused_bias.py [hunk 3]
```diff
@@ -118,3 +109,52 @@ def reset_parameters(self, prev_weight=None):
...
+
+
+def bias_dropout(
+    x: torch.Tensor,
+    bias: torch.Tensor,
+    p: float = 0.5,
+    training: bool = True,
+    inplace: bool = False,
+) -> torch.Tensor:
+    return F.dropout(x + bias, p=p, training=training, inplace=inplace)
```
Comment (szhengac) [inline slapo/op/fused_bias.py:121]: when inplace can be true?

File: slapo/op/fused_bias.py (hunk 2 above)
Comment (comaniac) [reply slapo/op/fused_bias.py:149]: This is a new API used in PyTorch 2.0. It applies AOTAutograd in advance (https://pytorch.org/functorch/stable/generated/functorch.compile.memory_efficient_fusion.html). The performance should be similar to TorchScript in our use cases.

File: slapo/op/fused_bias.py (hunk 3 above)
Comment (comaniac) [reply slapo/op/fused_bias.py:121]: I don't know actually. This is to align the signature of F.dropout.

File: slapo/op/fused_bias.py (hunk 2 above)
Comment (szhengac) [reply slapo/op/fused_bias.py:149]: "partitioner to perform efficient recomputation" means activation recomputation?

Comment (comaniac) [reply slapo/op/fused_bias.py:149]: It should refer to their algorithm (https://dev-discuss.pytorch.org/t/467) that is similar to our rematerialization pass in RAF. It uses a min-cut algorithm to selectively recompute some tensors. However, again since we only apply this API to a small module, it should not make any impact.

File: slapo/model_schedule/gpt2.py (hunk 1 above)
Comment (comaniac) [reply slapo/model_schedule/gpt2.py]: We have 2 options:
1. [Megatron] Apply .fork_rng() to top modules (embedding, transformers, pooler). The RNG by default uses the same seed within a TP group.
2. [Ours] The RNG has a mode to use different seeds within a TP group. With that, we don't need any .fork_rng() when sequence parallelism is enabled.

I was trying to use (1), but it turns out that it's tricky in Slapo, because `.fork_rng()` is s context manager, which cannot be traced. In other words, if we apply `sch.fork_rng()`, then we cannot trace the top module for pipeline partitioning. However, if we want to trace the top module and then apply `.fork_rng()`, we have to change the implementation in pipeline partitioning, because the current implementation of pipeline partitioning assume top modules are traced. Considering the efforts, I decided to go with (2) for now.

File: slapo/op/fused_bias.py (hunk 2 above)
Comment (comaniac) [reply slapo/op/fused_bias.py:149]: Yes. `memory_efficient_fusion` is a new API and should be getting mature in PT 2.0, so we may consider switching to that in the future.

File: slapo/op/fused_bias.py [hunk 4]
```diff
@@ -118,3 +109,52 @@ def reset_parameters(self, prev_weight=None):
...
+        prev_weight=None,
+    ):
+        factory_kwargs = {"device": device, "dtype": dtype}
+        super().__init__()
+        self.bias = torch.nn.Parameter(torch.empty(size, **factory_kwargs))
+        self.reset_parameters(prev_weight)
+        self.p = p
+        self.training = training
+        self.inplace = inplace
+        self.func = torch.jit.script(bias_dropout)
```
[the remark this answers is not in the document]
Comment (comaniac) [reply slapo/op/fused_bias.py:142]: No. The torchscript function can be used for training too. Otherwise the `.fuse` primitive won't work neither.

[the remark this answers is not in the document]
Comment (chhzh123) [reply slapo/op/fused_bias.py:142]: That is the trickiest part. If you only pass in a function to TorchScript, it will not optimize the backward pass. To let TorchScript capture the backward pass, the computation needs to be encapsulated in a `nn.Module`.

[the remark this answers is not in the document]
Comment (chhzh123) [reply slapo/op/fused_bias.py:142]: For `.fuse()` primitive, I always pass in a `GraphModule`, which is also a `nn.Module`, to TorchScript, so it is able to fuse the operations in the backward pass.

File: slapo/op/mlp.py [hunk 5]
```diff
@@ -23,12 +27,14 @@ def __init__(self, hidden_size, intermediate_size, orig_act, resid_pdrop):
         else:
             raise NotImplementedError(f"Unsupported activation: {orig_act}")
 
-        self.fc_out = torch.nn.Linear(intermediate_size, hidden_size)
-        self.dropout = torch.nn.Dropout(resid_pdrop)
+        self.fc_out = torch.nn.Linear(intermediate_size, hidden_size, bias=False)
+        self.bias_dropout = FusedBiasDropout(
+            hidden_size, resid_pdrop, prev_weight=self.fc_out.weight
+        )
```
[the remark this answers is not in the document]
Comment (comaniac) [reply slapo/op/mlp.py:33]: This is actually a good point and I also realized that in these days. I may try to refactor our ops to avoid this case as possible.

File: slapo/op/fused_bias.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (comaniac) [reply slapo/op/fused_bias.py:142]: I see. That's tricky. But the current bias_gelu is implemented in the same way. Does that mean it only affect the backward performance?

File: slapo/op/mlp.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (comaniac) [reply slapo/op/mlp.py:33]: I just found that it doesn't matter in FusedMLP, because when replacing an entire MLP, you will need to consider parameter conversion anyways, because even the name `fc_out` may not match the name in the original MLP. Assuming the original MLP has `fc_1` and `fc_2`, then we have
```
mlp.fc1.weight -> mlp.fc_in.weight
mlp.fc1.bias -> mlp.act.bias
mlp.fc2.weight
```

File: slapo/op/fused_bias.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (chhzh123) [reply slapo/op/fused_bias.py:142]: Nope. We have the [backward function](https://github.com/awslabs/slapo/blob/cd22329a4344187347e08a3355c89a01b9ca636f/slapo/op/bias_gelu.py#L31-L39) for `bias_gelu`. This should work. I remember there will be ~3% performance degradation if we only fuse the forward op of `bias_gelu`. I'm not sure how much will be affected for `bias_dropout`.

File: slapo/op/mlp.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (chhzh123) [reply slapo/op/mlp.py:33]: Parameter initialization is not based on the attribute name but is based on the module class. See the HF [`_init_weights`](https://github.com/huggingface/transformers/blob/main/src/transformers/models/bert/modeling_bert.py#L746-L760). This method is applied recursively, which means it does not need to a mapping convertion.

File: slapo/op/fused_bias.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (comaniac) [reply slapo/op/fused_bias.py:142]: Sorry I meant `bias_new_gelu`...so I guess this is another reason of using `memory_efficient_fusion`, because it will perform AOTAugograd and also optimize the backward graph.

File: slapo/op/mlp.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (comaniac) [reply slapo/op/mlp.py:33]: I see so you're talking about pre-training. I was referring to model checkpoints. That's a fair call then.

File: slapo/op/fused_bias.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (chhzh123) [reply slapo/op/fused_bias.py:142]: Make sense. But for now, we can encapsulate the function in a module as a workaround.

[the remark this answers is not in the document]
Comment (chhzh123) [reply slapo/op/fused_bias.py:142]: And yeah, I checked the implementation of `bias_new_gelu`. It indeed can lead to possible performance degradation. Do we have any models using `FusedBiasNewGELU`?

File: slapo/op/mlp.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (chhzh123) [reply slapo/op/mlp.py:33]: Okay, for common cases, I would expect module replacement does not change the original behavior. If the parameter values are also changed, it will be hard to verify or future debug.

File: slapo/op/fused_bias.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (comaniac) [reply slapo/op/fused_bias.py:142]: The default config of GPT2 and GPT-Neo use it.

[the remark this answers is not in the document]
Comment (comaniac) [reply slapo/op/fused_bias.py:142]: Oh I guess it might be ok because I used `memory_efficient_fusion` in BiasNewGeLU...

[the remark this answers is not in the document]
Comment (chhzh123) [reply slapo/op/fused_bias.py:142]: Emm. The 2% performance gap between Nemo-Megatron...

[the remark this answers is not in the document]
Comment (chhzh123) [reply slapo/op/fused_bias.py:142]: Did you use pytorch 2.0 when benchmarking the GPT2 model?

[the remark this answers is not in the document]
Comment (comaniac) [reply slapo/op/fused_bias.py:142]: This API is available in 1.13 already.

**Comment by comaniac** [comment]
Based on the comments, I removed the fused bias modules. Instead, I created `LinearWithAct` and `LinearWithDropout`. Both modules derive `nn.Linear`. In addition, I enable `memory_efficient_fusion` for bias+dropout if available, and add a flag to switch to torchscript in unit tests.

File: slapo/op/linear.py [hunk 6]
```diff
@@ -45,3 +65,118 @@ def forward(self, x: Tensor) -> Tensor:
         x = F.linear(x, self.weight, None)
         x = x + self.bias
         return x
+
+
+class LinearWithAct(nn.Linear):
+    """Derived from `nn.Linear` but with the following acvitation functions fused.
```
Comment (szhengac) [inline slapo/op/linear.py:71]: activation

**Review by szhengac — APPROVED** [review]

File: slapo/op/linear.py (hunk 6 above)
Comment (comaniac) [inline slapo/op/linear.py:71]:
```suggestion
    """Derived from `nn.Linear` but with the following activation functions fused.
```

**Comment by comaniac** [comment]
Thanks @chhzh123 @szhengac
